### PR TITLE
feat(rust, python): Improve cut and allow use in expressions

### PR DESCRIPTION
**Pull request by magarick**
Cut can work directly on a series and return a series and it also works in df expressions. It no longer requires sorting and on first glance appears faster as well. This PR also fixes a bug where you could get incorrect results if the breaks weren't passed in sorted. The code is shorter too, but I've kept around the old version for now in case people still want it.

This is a WIP as I could use some feedback on if I did things the right way, but the code does appear to work fine.

### CODE REVIEW

**Comment by magarick** [comment]
I need to fix up the tests but this does seem to work for the most part.
However, there's a problem with `over` expressions. If the global string cache isn't enabled, there won't be different categories for different groups when there should be. And if it is enabled, I get an error when trying to do `qcut`. Any help would be appreciated, since working correctly over groups is important.

**Comment by magarick** [comment]
I'm using the new style expressions but still have an issue with "over". Any ideas for how to work around this would be helpful. One way that seems reasonable, but probably not ideal is, is to let the function return a series containing Strings and then convert that to a categorical in the function that the expression dispatches. But this feels like a hack and I'd like to understand what's going on.

**Comment by magarick** [comment]
Ok, it more or less works now. The only caveat is that if you use `qcut` in an `over` expression without specifying labels and the quantiles are different per group it ends up just taking labels from one of the groups :-( I'm not sure how to resolve this.

File: polars/polars-lazy/polars-plan/src/dsl/function_expr/mod.rs [hunk 1]
```diff
@@ -14,6 +14,7 @@ mod clip;
 mod concat;
 mod correlation;
 mod cum;
+mod cut;
```
Comment (ritchie46) [inline polars/polars-lazy/polars-plan/src/dsl/function_expr/mod.rs:17]: Can you feature gate this functionality?

File: polars/polars-lazy/polars-plan/src/dsl/function_expr/mod.rs [hunk 2]
```diff
@@ -192,6 +193,17 @@ pub enum FunctionExpr {
         method: correlation::CorrelationMethod,
         ddof: u8,
     },
+    Cut {
```
Comment (ritchie46) [inline polars/polars-lazy/polars-plan/src/dsl/function_expr/mod.rs:196]: Can you feature gate this functionality?

File: polars/polars-lazy/polars-plan/src/dsl/function_expr/mod.rs [hunk 3]
```diff
@@ -192,6 +193,17 @@ pub enum FunctionExpr {
...
+        breaks: Vec<f64>,
+        labels: Option<Vec<String>>,
+        left_closed: bool,
+    },
+    QCut {
```
Comment (ritchie46) [inline polars/polars-lazy/polars-plan/src/dsl/function_expr/mod.rs:201]: Can you feature gate this functionality?

File: polars/polars-ops/src/series/ops/cut.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
+use std::iter::once;
+
+use polars_core::prelude::*;
+
+use crate::series::ops::SeriesSealed;
+
+pub trait CutSeries: SeriesSealed {
```
Comment (ritchie46) [inline polars/polars-ops/src/series/ops/cut.rs:7]: We only dispatch via expression. Can make it simpler by just having a function that accepts a `&Series` and arguments?

**Comment by magarick** [comment]
I feature gated and simplified dispatch. Less code is better code.
HOWEVER, I now have the same problem as when I used the string cache manually:
In the following example, everything but the last line works
```python
import polars as pl
x = pl.Series(range(20))
r = pl.Series([pl.repeat('a', 10, eager = True), pl.repeat('b', 10, eager = True)]).explode()
df = pl.DataFrame(dict(x = x, g = r))

x.qcut([.5], series = True)

df.with_columns(pl.col('x').qcut([.5], labels = ["less", "more"]).over('g'))

df.with_columns(pl.col('x').qcut([.5]).over('g'))```

The error is
```thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /Users/josh/repos/polars/polars/polars-core/src/chunked_array/logical/categorical/builder.rs:107:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/josh/repos/polars/py-polars/polars/dataframe/frame.py", line 1397, in __repr__
    return self.__str__()
  File "/Users/josh/repos/polars/py-polars/polars/dataframe/frame.py", line 1394, in __str__
    return self._df.as_str()
pyo3_runtime.PanicException: called `Option::unwrap()` on a `None` value```

I don't know if I'm building the categorical incorrectly or if I've misunderstood something in the code. It works when sharing labels, and it works when not in an over expression.

**Comment by magarick** [comment]
The problem seems to be related to how `over` maps values back to their original row. If I understand correctly, window functions first perform an aggregation, giving you a `Series` with a `list` for each group and then explodes and reshuffles the results. I've recreated it in a simpler form.

```python
>>> pl.enable_string_cache(False)
>>> s1 = pl.Series(['a', 'b'], dtype=pl.Categorical)
>>> s2 = pl.Series(['c', 'd'], dtype=pl.Categorical)
>>> pl.DataFrame(dict(c = [s1, s2]))
shape: (2, 1)
┌────────────┐
│ c          │
│ ---        │
│ list[cat]  │
╞════════════╡
│ ["a", "b"] │
│ ["a", "b"] │
└────────────┘
>>>
>>> pl.enable_string_cache(True)
>>> s1 = pl.Series(['a', 'b'], dtype=pl.Categorical)
>>> s2 = pl.Series(['c', 'd'], dtype=pl.Categorical)
>>> pl.DataFrame(dict(c = [s1, s2]))
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /Users/josh/repos/polars/polars/polars-core/src/chunked_array/logical/categorical/builder.rs:107:42
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/josh/repos/polars/py-polars/polars/dataframe/frame.py", line 1397, in __repr__
    return self.__str__()
  File "/Users/josh/repos/polars/py-polars/polars/dataframe/frame.py", line 1394, in __str__
    return self._df.as_str()
pyo3_runtime.PanicException: called `Option::unwrap()` on a `None` value
```

In the first case, levels get silently mis-mapped. The second has the same error as the `qcut` example.

**Comment by magarick** [comment]
In case it helps, I get the same error if I try `pl.Series([s1, s2])`.
And it seems like _maybe_ this has something to do with `NamedFrom`, specifically for `AsRef<[Series]>`? and how it uses `get_list_builder`. I don't see any handling for categoricals or a `ListCategoricalChunkedBuilder` like with other data types. I don't know if this is even related but maybe? @ritchie46 any help would be appreciated since this seems to be getting deep into internals. If you're not sure I'll try again in the morning. Thanks.

**Comment by ritchie46** [comment]
Ok, I will see if I can fix the `Series([s1, s2])` bug. 

For this PR, I think we should raise if qcut is used in window expressions. There is a lot of complexity involved and I don't see how we can bubble up the data type properly in window functions. 

Let's raise until we are a bit further in the core architecture and then we can maybe deal with it.

**Comment by ritchie46** [comment]
Thanks @magarick. I hope to lift the restrictions regarding groupby/window functions soon!

**Pull request merged**